### ISSUE TITLE
How to create a Splash screen?

### ISSUE THREAD

**Issue by msasso69** [picoe/Eto#926]
I apologize for my very basic question, but what is the best way to have a splash screen that updates while the main form is loading?

Thanks

**Comment by msasso69** [comment]
A second form from the same UI thread does the job! How silly of me :(

**Comment by msasso69** [comment]
At the end I ended up with a code similar to the following:

	using System;
	using Eto;
	using Eto.Forms;
	using Eto.Drawing;

	namespace VisualSEO.EtoFormGui
	{
		public class Program
		{
			[STAThread]
			public static void Main(string[] args)
			{
				Application app = new Application(Platform.Detect);
				Form form = new MySplash(app);
				app.Run(form);
			}
		}

		public class MySplash : Form
		{
			Label label;
			UITimer timer;
			int seconds;
			Application app;

			public MySplash(Application app)
			{
				this.app = app;
				WindowStyle = WindowStyle.None;
				Topmost = true;

				// size the splash
				Size = new Size(500, 300);
				// position it at the center of the screen
				Location = new Point((int)(Screen.WorkingArea.Width - Size.Width) / 2, (int)(Screen.WorkingArea.Height - Size.Height) / 2);
				Content = new StackLayout
				{
					VerticalContentAlignment = VerticalAlignment.Center,
					HorizontalContentAlignment = HorizontalAlignment.Center,
					Items =
					{
						null,
						new StackLayoutItem(label = new Label()),
						null
					}
				};
			}

			protected override void OnShown(EventArgs e)
			{
				base.OnShown(e);

				timer = new UITimer();
				timer.Interval = 1;
				timer.Elapsed += (sender, ev) =>
				{
					label.Text = seconds.ToString();
					if (seconds == 5)
					{
						Form mainform = new MyMainForm();
						app.MainForm = mainform;
						mainform.Show();
					}
					if (seconds == 10)
					{
						timer.Stop();
						this.Close(); // on Wpf works only with TomQv patch (#816)
					}
					seconds++;
				};
				timer.Start();
			}
		}

		public class MyMainForm : Form
		{
			public MyMainForm()
			{
				WindowState = WindowState.Maximized;
			}
		}
	}

Unfortunately on wpf the splash still keeps a little border and can be resized (I'm running it on Win7, I haven't tried on Win10). How can it be avoided?

![image](https://user-images.githubusercontent.com/21080030/33372727-4db7c4fa-d4ff-11e7-946d-140de17eed53.png)

Note that in order to close the splash I had to apply the patch  in issue #816  (thanks TomQv)

**Comment by msasso69** [comment]
I tried this other approch creating and showing both the splash and the main form in the Application.Initialized (Inspired by issue #416) and simulating a long processing afterward, but the risult is the same: the label in the splash is never refreshed apart from the very end and I do not really understand why.

	using System;
	using Eto;
	using Eto.Forms;
	using Eto.Drawing;
	using System.Threading;

	namespace VisualSEO.EtoFormGui
	{
		public class Program
		{
			[STAThread]
			public static void Main(string[] args)
			{
				Application app = new Application(Platform.Detect);
				app.Initialized += (sender, e) =>
				{
					MySplash splash = new MySplash();
					MyMainForm mainform = new MyMainForm(splash);
					splash.MainForm = mainform;
					app.MainForm = mainform;
					splash.Show();
					mainform.Show();
				};
				app.Run();
			}
		}

		public class MySplash : Form
		{
			Label label;
			UITimer timer;
			int counter;

			public MySplash()
			{
				WindowStyle = WindowStyle.None;
				Topmost = true;

				// size the splash
				Size = new Size(500, 300);
				// position it at the center of the screen
				Location = new Point((int)(Screen.WorkingArea.Width - Size.Width) / 2, (int)(Screen.WorkingArea.Height - Size.Height) / 2);
				Content = new StackLayout
				{
					VerticalContentAlignment = VerticalAlignment.Center,
					HorizontalContentAlignment = HorizontalAlignment.Center,
					Items =
						{
							null,
							new StackLayoutItem(label = new Label()),
							null
						}
				};
			}

			protected override void OnShown(EventArgs e)
			{
				base.OnShown(e);

				MainForm.SimulateWork();

				timer = new UITimer();
				timer.Interval = 1;
				timer.Elapsed += (sender, ev) =>
				{
					if (MainForm != null && MainForm.WorkCompleted)
					{
						timer.Stop();
						this.Close();
					}
				};
				timer.Start();
			}

			public void IncrementCounter()
			{
				counter++;
				label.Text = counter.ToString();
				label.Invalidate();
			}

			public MyMainForm MainForm { get; set; }
		}

		public class MyMainForm : Form
		{
			MySplash splash;

			public MyMainForm(MySplash splash)
			{
				this.splash = splash;
				WindowState = WindowState.Maximized;
			}

			public void SimulateWork()
			{
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				WorkCompleted = true;
			}

			public bool WorkCompleted { get; set; }
		}
	}

Basically I'm still stuck in managing to have a splash screen that updates on the progress of the initialization of the main form :(

**Comment by SlowLogicBoy** [comment]
Hey, this is all I needed for borders to dissapear:
```csharp
new Application().Run(new Form {
    WindowStyle = WindowStyle.None,
    Resizable = false
});
```

**Comment by msasso69** [comment]
Hi @SlowLogicBoy, 

Resizable = false makes the border disappear, thanks a lot! :)

Your suggestion for the label increment however has no effect :(

Temporarily I've found a very ugly trick that does the magic: if after updating the label I pop up a form (on Wpf) or a dialog (on Mac) the label gets updated, so I show and immediately close a small empty form/dialog behind the splash and I'm done. But as I said it's really an ugly trick that I hope I'll remove soon.

**Comment by SlowLogicBoy** [comment]
Well, label doesn't update most likely because UI thread gets locked, that's why I put those sleep actions in Task.Run... For you it might not work because I didn't put `counter++` to UI thread too.

**Comment by msasso69** [comment]
Sorry, my mistake, I forgot to add the Task.Run() part. Your code works, thanks a lot! :)
I now need to apply it in my real code but that is my problem. Thanks again!
(I think that this issue can be closed now)

**Comment by msasso69** [comment]
One more thing (sorry if I'm getting boring):

In my real application even if I moved lots of logic outside the constructor of main form the constructor itself still takes a lot of time to complete. Back to the given example if MyMainForm takes a lot of time and since both MySplash and MyMainForm are created and shown together in Application.Instanziated they are also shown at the same time but only after the constructor of MyMainForm has finished to create the form.

It becomes evident just adding a sleep in the constructor of MyMainForm like this:

		public MyMainForm(MySplash splash)
		{
			this.splash = splash;
			WindowState = WindowState.Maximized;
			System.Threading.Thread.Sleep(5000);
		}

My desiderata is to see the splash screen right after I launch the application and make it disappear when everything is fully created and initialized.

I therefore changed my example a little bit in order to create MyMainForm only after the Splash is shown
So I removed the creation of MyMainForm from Application.Instanziated and placed the following code in the MySplash.OnShown

		protected override void OnShown(EventArgs e)
		{
			base.OnShown(e);

			Thread thread = new Thread(new ThreadStart(CreateMyMainForm));
			thread.SetApartmentState(ApartmentState.STA);
			thread.Start();

where CreateMyMainForm() is:

			void CreateMyMainForm()
			{
				MyMainForm mainform = new MyMainForm(this);
				this.MainForm = mainform;
				Application.Instance.AsyncInvoke(() =>
				{
					Application.Instance.MainForm = mainform; // this should ensure that closing MySplash is not closing the whole application
				});
				mainform.Show();
				mainform.SimulateWork();
			}

Now everything seems to work apart from the fact that when I close the splash screen the whole application closes (or crashes, it's not clear). To be precise on Mac the application exits, on Wpf all forms close and the application remains running in background (note that on Wpf I applied TomQv's patch #816). 

What am I doing wrong?

Here is the full code:

	using System;
	using Eto;
	using Eto.Forms;
	using Eto.Drawing;
	using System.Threading;
	using System.Threading.Tasks;

	namespace VisualSEO.EtoFormGui
	{
		public class Program
		{
			[STAThread]
			public static void Main(string[] args)
			{
				//new Application(Platform.Detect).Run(new MySplash());

				Application app = new Application(Platform.Detect);
				app.Initialized += (sender, e) =>
				{
					MySplash splash = new MySplash();
					app.MainForm = splash;
					splash.Show();
				};
				app.Run();
			}
		}

		public class MySplash : Form
		{
			Label label;
			UITimer timer;
			string text;
			int counter;

			public MySplash()
			{
				BackgroundColor = Colors.LightYellow;
				ShowInTaskbar = false;
				WindowStyle = WindowStyle.None;
				Resizable = false;
				Topmost = true;

				// size the splash
				Size = new Size(500, 300);
				// position it at the center of the screen
				Location = new Point((int)(Screen.WorkingArea.Width - Size.Width) / 2, (int)(Screen.WorkingArea.Height - Size.Height) / 2);
				Content = new StackLayout
				{
					VerticalContentAlignment = VerticalAlignment.Center,
					HorizontalContentAlignment = HorizontalAlignment.Center,
					Items =
					{
						null,
						new StackLayoutItem(label = new Label()),
						null
					}
				};
			}

			void CreateMyMainForm()
			{
				MyMainForm mainform = new MyMainForm(this);
				this.MainForm = mainform;
				Application.Instance.AsyncInvoke(() =>
				{
					Application.Instance.MainForm = mainform; // this should ensure that closing MySplash is not closing the whole application
				});
				mainform.Show();
				mainform.SimulateWork();
			}

			protected override void OnShown(EventArgs e)
			{
				base.OnShown(e);

				Thread thread = new Thread(new ThreadStart(CreateMyMainForm));
				thread.SetApartmentState(ApartmentState.STA);
				thread.Start();

				timer = new UITimer();
				timer.Interval = 1;
				timer.Elapsed += (sender, ev) =>
				{
					if (MainForm != null && MainForm.WorkCompleted)
					{
						timer.Stop();
						this.Close(); // this is closing averything instead of just MySplash
					}
				};
				timer.Start();
			}

			public void SetText(string text)
			{
				this.text = text;
			}

			public void IncrementCounter()
			{
				counter++;
				Application.Instance.AsyncInvoke(() =>
				{
					label.Text = string.Format("{0} {1}", text, counter.ToString());
				});
			}

			public MyMainForm MainForm { get; set; }
		}
		
		public class MyMainForm : Form
		{
			MySplash splash;

			public MyMainForm(MySplash splash)
			{
				this.splash = splash;
				WindowState = WindowState.Maximized;

				splash.SetText("MyMainForm()");
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
			}

			public void SimulateWork()
			{
				splash.SetText("SimulateWork()");
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();
				System.Threading.Thread.Sleep(1000);
				splash.IncrementCounter();

				WorkCompleted = true;
			}

			public bool WorkCompleted { get; set; }
		}
	}

**Comment by SlowLogicBoy** [comment]
There are quite a few problems here:

- In Form you should only create Controls that form uses and maybe some styling, nothing more nothing less. Right now you block UI thread with heavy jobs that's why they do not update :).
- All Jobs must be done in different classes, and update Form using Control binding (MVVM Design pattern) how to bind you can find [here](https://github.com/picoe/Eto/wiki/Data-Binding).

I solve this by creating services that do heavy lifting in a different thread and they communicate with ViewModel, through Messaging pattern or Event aggregator pattern.

**Comment by msasso69** [comment]
Ok, thank you @SlowLogicBoy, I'll try to follow your advise :)

**Comment by cra0zy** [comment]
@cwensley This should be closed, not a bug or a feature request.
